mlir/lib/Catalyst/Transforms/catalyst_to_llvm.cpp

```diff
@@ -479,6 +480,84 @@ struct PythonCallOpPattern : public OpConversionPattern<PythonCallOp> {
     }
 };
 
+struct CallbackOpPatternOne : public OpConversionPattern<CallbackOp> {
+    using OpConversionPattern::OpConversionPattern;
+
+    LogicalResult match(CallbackOp op) const override
+    {
+        // Only match with ops without an entry block
+        return !op.empty() ? failure() : success();
+    }
+
+    void rewrite(CallbackOp op, CallbackOpAdaptor adaptor,
+                 ConversionPatternRewriter &rewriter) const override
+    {
+        Block *entry;
+        rewriter.updateRootInPlace(op, [&] { entry = op.addEntryBlock(); });
+        PatternRewriter::InsertionGuard guard(rewriter);
+        rewriter.setInsertionPointToStart(entry);
+
+        auto ctx = rewriter.getContext();
+        Type ptrTy = LLVM::LLVMPointerType::get(ctx);
+        auto one = rewriter.getI64IntegerAttr(1);
+        auto loc = op.getLoc();
+        Value c1 = rewriter.create<LLVM::ConstantOp>(loc, one);
+        auto idAttr = op.getIdAttr();
+        auto constantId = rewriter.create<LLVM::ConstantOp>(loc, idAttr);
+        auto argcAttr = op.getArgcAttr();
+        auto constantArgc = rewriter.create<LLVM::ConstantOp>(loc, argcAttr);
+        auto rescAttr = op.getRescAttr();
+        auto constantResc = rewriter.create<LLVM::ConstantOp>(loc, rescAttr);
+
+        SmallVector<Value> callArgs = {constantId, constantArgc, constantResc};
+
+        Type i64 = rewriter.getI64Type();
+        Type voidType = LLVM::LLVMVoidType::get(ctx);
+        bool isVarArg = true;
+        ModuleOp mod = op->getParentOfType<ModuleOp>();
+        auto typeConverter = getTypeConverter();
+        LLVM::LLVMFuncOp customCallFnOp =
+            mlir::LLVM::lookupOrCreateFn(mod, "inactive_callback", {/*args=*/i64, i64, i64},
+                                         /*ret_type=*/voidType, isVarArg);
+
+        for (auto arg : llvm::enumerate(op.getArguments())) {
+            auto val = arg.value();
+            Type structTy = typeConverter->convertType(val.getType());
+            auto structVal =
+                rewriter.create<UnrealizedConversionCastOp>(loc, structTy, val).getResult(0);
```

Comment:
Unrealized cast to a pointer

Reply:
This one can't be removed. This is because the error is actually in wrapMemrefArgsFunc. It will introduce another unrealized conversion cast that converts the uses of the arguments to memref. So this means that `val` here needs to be converted from memref to structTy and back to a pointer if we want to reuse wrapMemrefArgsFunc.